File: frontend/src/auth/rbac.ts
import type { ErpNavGroup } from "../navigation/erpNavConfig";

/** Stored under `ERP_ROLE_STORAGE_KEY`; empty / missing = unrestricted (legacy behaviour). */
export type ErpRole =
  | "CEO"
  | "Obchod"
  | "Plánování"
  | "Výroba"
  | "Sklad"
  | "Kvalita"
  | "Technologie"
  | "Administrativa";

export const ERP_ROLE_STORAGE_KEY = "akeng_role";

export const ERP_ROLE_OPTIONS: { value: ErpRole; label: string }[] = [
  { value: "CEO", label: "CEO" },
  { value: "Obchod", label: "Obchod" },
  { value: "Plánování", label: "Plánování" },
  { value: "Výroba", label: "Výroba" },
  { value: "Sklad", label: "Sklad" },
  { value: "Kvalita", label: "Kvalita" },
  { value: "Technologie", label: "Technologie" },
  { value: "Administrativa", label: "Administrativa" },
];

const FULL_ACCESS = new Set<ErpRole>(["CEO", "Administrativa"]);

const NAV_BY_ROLE: Record<ErpRole, Set<string>> = {
  CEO: new Set(),
  Obchod: new Set(["dashboard", "orders", "purchase", "master_data"]),
  Plánování: new Set(["dashboard", "orders", "planning", "production", "warehouse", "master_data"]),
  Výroba: new Set(["dashboard", "orders", "production", "master_data"]),
  Sklad: new Set(["dashboard", "orders", "warehouse", "master_data"]),
  Kvalita: new Set(["dashboard", "orders", "production", "warehouse", "quality", "master_data"]),
  Technologie: new Set(["dashboard", "orders", "production", "technology", "master_data"]),
  Administrativa: new Set(),
};

/** Roles allowed per action (CEO / Administrativa / missing role bypass in UI). */
const ACTION_ROLES: Record<string, Set<ErpRole>> = {
  "orders.write": new Set(["Obchod"]),
  "orders.storno": new Set(),
  "planning.write": new Set(["Plánování"]),
  "production.execute": new Set(["Plánování", "Výroba"]),
  "production.storno": new Set(["Plánování"]),
  "stock.mutate": new Set(["Sklad"]),
  "technology.write": new Set(["Technologie"]),
  "quality.write": new Set(["Kvalita"]),
  "purchase.write": new Set(["Obchod"]),
};

export function readStoredErpRole(): ErpRole | null {
  try {
    const raw = localStorage.getItem(ERP_ROLE_STORAGE_KEY)?.trim();
    if (!raw) return null;
    const hit = ERP_ROLE_OPTIONS.find((o) => o.value === raw);
    return hit ? hit.value : null;
  } catch {
    return null;
  }
}

export function writeStoredErpRole(role: ErpRole | null): void {
  try {
    if (!role) localStorage.removeItem(ERP_ROLE_STORAGE_KEY);
    else localStorage.setItem(ERP_ROLE_STORAGE_KEY, role);
  } catch {
    /* ignore */
  }
}

export function filterNavGroupsByRole(role: ErpRole | null, groups: ErpNavGroup[]): ErpNavGroup[] {
  if (role == null || FULL_ACCESS.has(role)) return groups;
  const allowed = NAV_BY_ROLE[role];
  if (!allowed || allowed.size === 0) return groups;
  return groups.filter((g) => allowed.has(g.id));
}

export function canPerformAction(role: ErpRole | null, action: string): boolean {
  if (role == null || FULL_ACCESS.has(role)) return true;
  const set = ACTION_ROLES[action];
  if (!set) return true;
  return set.has(role);
}

// ---------------------------------------------------------------------------
// Permission-based gating (nová cesta — navázáno na DB knihovnu uživatelů).
//
// Frontend si drží cache aktuálního uživatele (`MeDto`) získaného z `/users/me`.
// `hasPermission(code)` čte z této cache. Pokud cache není naplněná, chováme se
// defenzivně: vrátíme true (pilot režim) stejně jako legacy `canPerformAction`,
// aby se UI nezablokovalo před dokončením prvního fetch /users/me.
// ---------------------------------------------------------------------------

export type CurrentUserSnapshot = {
  permissions: Set<string>;
  roles: Set<string>;
  hasFullAccess: boolean;
  username: string | null;
  displayName: string | null;
  loaded: boolean;
};

let CURRENT_USER: CurrentUserSnapshot = {
  permissions: new Set(),
  roles: new Set(),
  hasFullAccess: true, // default allow before /users/me loads — legacy pilot chování
  username: null,
  displayName: null,
  loaded: false,
};

type Listener = (snapshot: CurrentUserSnapshot) => void;
const LISTENERS: Set<Listener> = new Set();

export function getCurrentUserSnapshot(): CurrentUserSnapshot {
  return CURRENT_USER;
}

export function setCurrentUserSnapshot(next: Partial<CurrentUserSnapshot>): void {
  CURRENT_USER = {
    ...CURRENT_USER,
    ...next,
    permissions: next.permissions ?? CURRENT_USER.permissions,
    roles: next.roles ?? CURRENT_USER.roles,
    loaded: next.loaded ?? true,
  };
  LISTENERS.forEach((fn) => {
    try {
      fn(CURRENT_USER);
    } catch {
      /* ignore */
    }
  });
}

export function subscribeCurrentUser(listener: Listener): () => void {
  LISTENERS.add(listener);
  return () => {
    LISTENERS.delete(listener);
  };
}

export function hasPermission(code: string): boolean {
  const snap = CURRENT_USER;
  if (!snap.loaded) return true; // pilot fallback dokud se /users/me nenačte
  if (snap.hasFullAccess) return true;
  return snap.permissions.has(code);
}

export function hasAnyPermission(codes: string[]): boolean {
  if (codes.length === 0) return true;
  return codes.some((c) => hasPermission(c));
}

/** Vhodné pro guard komponentu `<PermissionGate permission="manage_users">`. */
export function canSeeUsersLibrary(): boolean {
  return hasPermission("manage_users");
}

File: backend/app/core/rbac.py
"""
Simple role-based access for AKENG ERP.

- Missing / unknown role header → allow all (pilot / backward compatible).
- CEO and Administrativa → full access when a valid role is supplied.
"""

from __future__ import annotations

import unicodedata

from fastapi import HTTPException


def _fold_ascii(s: str) -> str:
    return "".join(
        c for c in unicodedata.normalize("NFD", s.lower()) if unicodedata.category(c) != "Mn"
    )

# Canonical labels (UI / header), Czech diacritics as in product spec.
ROLES: tuple[str, ...] = (
    "CEO",
    "Obchod",
    "Plánování",
    "Výroba",
    "Sklad",
    "Kvalita",
    "Technologie",
    "Administrativa",
)

_ROLE_ALIASES: dict[str, str] = {
    "ceo": "CEO",
    "obchod": "Obchod",
    "planovani": "Plánování",
    "planning": "Plánování",
    "vyroba": "Výroba",
    "production": "Výroba",
    "sklad": "Sklad",
    "warehouse": "Sklad",
    "kvalita": "Kvalita",
    "quality": "Kvalita",
    "technologie": "Technologie",
    "technology": "Technologie",
    "administrativa": "Administrativa",
    "admin": "Administrativa",
}

# Nav group ids — must match frontend `erpNavConfig` `ErpNavGroup.id`.
_NAV_BY_ROLE: dict[str, frozenset[str]] = {
    "Obchod": frozenset({"dashboard", "orders", "purchase", "master_data"}),
    "Plánování": frozenset({"dashboard", "orders", "planning", "production", "warehouse", "master_data"}),
    "Výroba": frozenset({"dashboard", "orders", "production", "master_data"}),
    "Sklad": frozenset({"dashboard", "orders", "warehouse", "master_data"}),
    "Kvalita": frozenset({"dashboard", "orders", "production", "warehouse", "quality", "master_data"}),
    "Technologie": frozenset({"dashboard", "orders", "production", "technology", "master_data"}),
}

# Action → roles allowed (CEO / Administrativa handled separately).
_ACTION_ROLES: dict[str, frozenset[str]] = {
    "orders.write": frozenset({"Obchod"}),
    "orders.storno": frozenset({}),  # CEO / Administrativa only
    "planning.write": frozenset({"Plánování"}),
    "production.execute": frozenset({"Plánování", "Výroba"}),
    "production.storno": frozenset({"Plánování"}),
    "stock.mutate": frozenset({"Sklad"}),
    "technology.write": frozenset({"Technologie"}),
    "quality.write": frozenset({"Kvalita"}),
    "purchase.write": frozenset({"Obchod"}),
}


def normalize_role(raw: str | None) -> str | None:
    if raw is None:
        return None
    s = str(raw).strip()
    if not s:
        return None
    if s in ROLES:
        return s
    return _ROLE_ALIASES.get(_fold_ascii(s))


def is_full_access_role(role: str | None) -> bool:
    return role in ("CEO", "Administrativa")


def can_see_nav_group(role: str | None, group_id: str) -> bool:
    if role is None or is_full_access_role(role):
        return True
    allowed = _NAV_BY_ROLE.get(role)
    if allowed is None:
        return True
    return group_id in allowed


def can_perform_action(role: str | None, action: str) -> bool:
    if role is None:
        return True
    if is_full_access_role(role):
        return True
    allowed_roles = _ACTION_ROLES.get(action)
    if allowed_roles is None:
        return True
    return role in allowed_roles


def assert_can(role: str | None, action: str) -> None:
    if can_perform_action(role, action):
        return
    raise HTTPException(status_code=403, detail="Nedostatečná oprávnění pro tuto akci.")
